packages/core/src/scuttle.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * @typedef {object} ScuttleOpts
+ * @property {boolean} enabled - Whether scuttling is enabled or not.
+ * @property {Array<string|RegExp>} exceptions - List of properties to exclude from scuttling.
+ * @property {string} scuttlerName - Name of the scuttler function to use which is expected to be found as a
+ * property on the global object (e.g. if scuttlerName is 'x', scuttler function is obtained from globalThis['x']).
+ */
+
+/**
+ * @typedef {object} GlobalRef
+ * @property {Record<string, any>} [globalThis] - Reference to the global object.
+ */
+
+const { Object, Array, Error, RegExp, Set, console, Proxy, Reflect } =
+  globalThis
+
+const {
+  assign,
+  getOwnPropertyNames,
+  getOwnPropertyDescriptor,
+  create,
+  defineProperty,
+} = Object
+
+const { isArray, from } = Array
+
+const { getPrototypeOf } = Reflect
+
+const { warn } = console
+
+function generateInvokers(prop) {
+  return { get, set }
+  function set() {
+    warn(
+      `LavaMoat - property "${prop}" of globalThis cannot be set under scuttling mode. ` +
+        'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+    )
+  }
+  function get() {
+    throw new Error(
+      `LavaMoat - property "${prop}" of globalThis is inaccessible under scuttling mode. ` +
+        'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+    )
+  }
+}
+
+/**
+ * Applies scuttling, with the default set of options, including using Snow if passed in as scuttlerFunc.
+ * Scuttle globalThis right after we used it to create the root package compartment.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts} opts - Scuttling options.
+ */
+function scuttle(globalRef, opts) {
+  const scuttleOpts = generateScuttleOpts(globalRef, opts)
+
+  if (scuttleOpts.enabled) {
+    if (!isArray(scuttleOpts.exceptions)) {
+      throw new Error(
+        `LavaMoat - exceptions must be an array, got ${typeof scuttleOpts.exceptions}`
+      )
+    }
+    scuttleOpts.scuttlerFunc(globalRef, (realm) =>
+      performScuttleGlobalThis(realm, scuttleOpts.exceptions)
+    )
+  }
+}
+
+/**
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts|boolean} originalOpts - Scuttling options. Accepts `true` for backwards compatibility.
+ * @returns {ScuttleOpts} - Final scuttling options.
+ */
+function generateScuttleOpts(globalRef, originalOpts = create(null)) {
+  const defaultOpts = {
+    enabled: true,
+    exceptions: [],
+    scuttlerName: '',
+  }
+  const opts = assign(
+    {},
+    originalOpts === true ? { ...defaultOpts } : { ...originalOpts },
```

Comment:
For my education: What is the significance of the spread here? This would seem equivalent to me but I guess there may be cases where it's not?
```suggestion
    originalOpts === true ? defaultOpts : originalOpts,
```
, as `assign` already limits its behavior to enumerable own props and only the value in the first parameter is modified.

Reply:
yeah, seems like you could just skip the `...` since the `defaultOpts` are not created outside of the function scope, so you don't need to worry about mutating it

Reply:
086ed7ac0